### PR TITLE
animdl: init at 1.7.27

### PR DESCRIPTION
**Pull request by PassiveLemon**
###### Description of changes
Adds Animdl, a cli scraper/downloader/streamer.
Adds Anchor, a Python library for scraping.
Adds Anitopy, a Python library for parsing anime video filenames.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PassiveLemon** [comment]
For everyone who has been trying to review my PRs, I am sorry, this has been chaotic

**Comment by PassiveLemon** [comment]
@Janik-Haag Also sorry for my PRs, they have been a little chaotic. I think I got it all sorted out though

**Comment by PassiveLemon** [comment]
@Janik-Haag Also sorry for my PRs, they have been a little chaotic. I think I got it all sorted out though

**Comment by PassiveLemon** [comment]
Huh? I did not change any endings. It's LF like it should be...

**Review by NobbZ — CHANGES_REQUESTED** [review]
Please squash these commits into a single one:

* 8dbe058dad101a1320e9ff0ac0b5e77c143d1c54
* 20cb3797eb0b3da3f559b4cbcd5cb0d3d0578244

**Comment by NobbZ** [comment]
As this is actually a python package, I'd like to also request the review of the python team, which I can not tag directly and will therefore ping @mweinelt in person, as he is the only python team member I know from the top of my head…

**Comment by PassiveLemon** [comment]
Finally. Not sure what that last check means

File: pkgs/applications/video/animdl/default.nix
Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:2]: This should use `python3.pkgs.buildPythonApplication` instead of poetry2nix.

File: pkgs/applications/video/animdl/default.nix
Comment (mweinelt) [inline pkgs/applications/video/animdl/default.nix:19]: These packages need individual packaging.

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:2]: The Animdl build instructions said to use poetry so I just used this as it seemed the easiest. I can try doing it that way if needed

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:19]: How would I do that?

Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:19]: Visit https://nixos.org/manual/nixpkgs/unstable/ and search for

> Python library packages in Nixpkgs

File: pkgs/applications/video/animdl/default.nix
Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:2]: Visit [nixos.org/manual/nixpkgs/unstable](https://nixos.org/manual/nixpkgs/unstable/) and search for

> buildPythonApplication function

**Comment by PassiveLemon** [comment]
Upstream is not tagging their release.
https://github.com/justfoolingaround/animdl/issues/277

File: pkgs/applications/video/animdl/default.nix
Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:2]: Set `format = "pyproject"` and put `poetry-core` into `nativeBuildInputs`. The python builder does support that.

Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:2]: I've got pycryptodomex in the attribute set and in the nativeBuildInputs but it complains that it could not find a version that is satisfactory: pycryptodomex<3.15.0,>=3.14.1.
I tried overwriting but that did not seem to work.

Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:2]: Try relaxing that requirement in `postPatch` using `substituteInPlace`. Search `pkgs/development/python-modules` for examples-

Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:2]: I have it successfully built. I had to pack 2 of its python dependencies so I have a question: What do I do with those? Do I pack them with this or make a PR for those?

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:19]: Done. Read the other comment I posted

File: pkgs/applications/video/animdl/default.nix
Comment (mweinelt) [reply pkgs/applications/video/animdl/default.nix:2]: They go individually into `pkgs/development/python-modules/` and are referenced from `pkgs/top-level/python-modules.nix`.

1. Yes, one commit per attribute
2. Bug the upstream to add one, otherwise it would need to be marked unfree

**Comment by PassiveLemon** [comment]
How's it now?

File: pkgs/applications/video/animdl/default.nix
Comment (NobbZ) [reply pkgs/applications/video/animdl/default.nix:24]: Please put a comment explaining why you use a commit-SHA directly.

Please also put a link to the issue asking for tags in that comment.

**Review by NobbZ — CHANGES_REQUESTED** [review]
Please make sure:

* You have on commit for adding yourself to the maintainers
* one commit per "package" created
* Having each of them named and described appropriately as explained in CONTRIBUTING.md

**Comment by PassiveLemon** [comment]
How do you reckon I fix this? I can't squash them

**Comment by mweinelt** [comment]
You could soft-reset onto the base of your pull request and create new commits from there.

```console
❯ git merge-base pr/240329 origin/master
3a4e234b072bd69e8901b8ed465be7bb659d1a93
❯ git reset --soft 3a4e234b072bd69e8901b8ed465be7bb659d1a93
❯ git status
On branch pr/240329
You are currently bisecting, started from branch '9e4e0807d214'.
  (use "git bisect reset" to get back to the original branch)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   maintainers/maintainer-list.nix
	new file:   pkgs/applications/video/animdl/default.nix
	new file:   pkgs/development/python-modules/anchor-kr/default.nix
	new file:   pkgs/development/python-modules/anitopy/default.nix
	modified:   pkgs/top-level/all-packages.nix
	modified:   pkgs/top-level/python-packages.nix
❯ git restore --staged .
❯ git status
On branch pr/240329
You are currently bisecting, started from branch '9e4e0807d214'.
  (use "git bisect reset" to get back to the original branch)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   maintainers/maintainer-list.nix
	modified:   pkgs/top-level/all-packages.nix
	modified:   pkgs/top-level/python-packages.nix

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	pkgs/applications/video/animdl/
	pkgs/development/python-modules/anchor-kr/
	pkgs/development/python-modules/anitopy/

no changes added to commit (use "git add" and/or "git commit -a")
```

From here you can `git add -p <path>` to add hunks to your staging area and `git commit` to create a commit out of the selected hunks.

A more advanced alternative would be an interactive rebase, but it is a bit more involved.

**Comment by PassiveLemon** [comment]
Upstream is not tagging their release.
https://github.com/justfoolingaround/animdl/issues/277
I use the commit SHA because there is no tag for me to get a release version from.

**Review by NobbZ — CHANGES_REQUESTED** [review]
Please make sure:

* You have on commit for adding yourself to the maintainers
* one commit per "package" created
* Having each of them named and described appropriately as explained in CONTRIBUTING.md
* Have a comment in your nix file that explains the hardcoded sha1 rather than a tag with a link to the upstream ticket

**Comment by PassiveLemon** [comment]
Do you mind if I just make a new branch for this? I'm trying to figure out how to remove all of the old commits and I cannot manage to do it...

**Comment by Janik-Haag** [comment]
you can do `git status` and then take the hash of the oldest one you want to remove and write `git reset $hash --mixed, there is multiple modes, you probably want to use mixed here.

**Comment by PassiveLemon** [comment]
There we go. Only the 4 needed commits.

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:30]:
```suggestion
  version = "1.7.17";
  format = "pyproject";

  src = fetchFromGitHub {
    owner = "justfoolingaround";
    repo = "${pname}";
    # Using the commit hash because upstream does not have releases. https://github.com/justfoolingaround/animdl/issues/277
    rev = "fd3277fb919c4b5e7ef42daba0bd001695a9c124";
    hash = "sha256-pgtrStorkIJJ3oIGgRjNWYwcQ5jUC8IS5iI9CyDJkzE=";
  };

  nativeBuildInputs = [
    poetry-core
    cssselect
  ];

  propagatedBuildInputs = [
    anchor-kr
    anitopy
    click
    httpx
    lxml
    packaging
    pkginfo
    pyyaml
    pycryptodomex
    regex
    rich
    tqdm
    yarl
  ];
```

just some formatting nitpicks

File: pkgs/development/python-modules/anitopy/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anitopy/default.nix:10]:
```suggestion

  src = fetchFromGitHub {
    owner = "igorcmoura";
    repo = "${pname}";
    rev = "v${version}";
    hash = "sha256-xXEf7AJKg7grDmkKfFuC4Fk6QYFJtezClyfA3vq8TfQ=";
  };
```

Comment (Janik-Haag) [reply pkgs/development/python-modules/anitopy/default.nix:10]: You are also missing the format attribute

File: pkgs/development/python-modules/anchor-kr/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anchor-kr/default.nix:5]: same as anitopy

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
you should prepend the `anchor-kr` and `anitopy` commit with `python3Packages.`
you should also add `pythonImportsCheck` to every module.

File: pkgs/development/python-modules/anitopy/default.nix
Comment (PassiveLemon) [reply pkgs/development/python-modules/anitopy/default.nix:10]: I'm sorry but what exactly is different in this?

Comment (Janik-Haag) [reply pkgs/development/python-modules/anitopy/default.nix:10]: There is just a space more. just a formating change.

Comment (Janik-Haag) [reply pkgs/development/python-modules/anitopy/default.nix:10]: and with the format attribute I mean something like `format = "setuptools";`

**Comment by PassiveLemon** [comment]
Updated with the requests.

**Review by Janik-Haag — COMMENTED** [review]
Can you do a rebase, should be good to go afterwards.

**Comment by PassiveLemon** [comment]
Trailing whitespace? This was never an issue before.

**Comment by Janik-Haag** [comment]
> Trailing whitespace? This was never an issue before.

Then you just didn't have any before ;)
CI still fails because of them: https://github.com/NixOS/nixpkgs/actions/runs/5648374159/job/15300460602?pr=240329

> pkgs/applications/video/animdl/default.nix:
>	32: Trailing whitespace
>
> 1 errors found
> Error: Process completed with exit code 123.

**Comment by PassiveLemon** [comment]
Not sure how that managed to get there but it should be all good now.

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:65]:
```suggestion
    maintainers = with maintainers; [ passivelemon ];
```
This has to be lowercase because it's defined in lower case in `maintainers/maintainer-list.nix` and afaik we only allow lower case maintainer entrys. Please change this for every package, after that it should be good to go. You can check if the package build fines localy if you execute `nix-build -A $pname` or python3Package.$pname etc.. in the root of your nixpkgs repo

Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix]: Done.

**Comment by PassiveLemon** [comment]
I'm not exactly sure how I'm supposed to fix this. I think it fails because of the new python libraries being added at the same time as animdl so it can't find them.

File: pkgs/applications/video/animdl/default.nix
Comment (eclairevoyant) [inline pkgs/applications/video/animdl/default.nix:27]:
```suggestion
    repo = "animdl";
```

File: pkgs/development/python-modules/anchor-kr/default.nix
Comment (eclairevoyant) [inline pkgs/development/python-modules/anchor-kr/default.nix:8]: this as well

File: pkgs/development/python-modules/anitopy/default.nix
Comment (eclairevoyant) [inline pkgs/development/python-modules/anitopy/default.nix:8]: and this

File: pkgs/applications/video/animdl/default.nix
Comment (eclairevoyant) [inline pkgs/applications/video/animdl/default.nix:65]: is this a binary? then we can add
```suggestion
    license = licenses.gpl3Only;
    mainProgram = "animdl";
```
or such

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (eclairevoyant) [inline pkgs/top-level/all-packages.nix:42286]: this should probably use python3's callPackage to build properly. did you test this?

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:27]: Would you mind explaining to me why? I've seen this in many other derivations

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:65]: This as well

File: pkgs/applications/video/animdl/default.nix
Comment (eclairevoyant) [reply pkgs/applications/video/animdl/default.nix:27]: it's bad practice and semanticlaly repo name is not the package name

File: pkgs/applications/video/animdl/default.nix
Comment (eclairevoyant) [reply pkgs/applications/video/animdl/default.nix:65]: what is your question?

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix]: Understood

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:66]: Just curious as to why I should add it?

Comment (eclairevoyant) [reply pkgs/applications/video/animdl/default.nix:66]: #246386

File: pkgs/top-level/all-packages.nix
Comment (PassiveLemon) [reply pkgs/top-level/all-packages.nix:42286]: Oh yeah, I have this locally, forgot to add it to the PR.

**Comment by PassiveLemon** [comment]
Updated with changes.

**Review by eclairevoyant — APPROVED** [review]

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:1]: python3Packages here is unused, you should remove it

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:21]: the `rec` here is also unused please remove it

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:33]: usually we put the format under the version attribute so like:
```
  pname = "animdl";
  version = "1.7.27";
  format = "pyproject";
```
so it would be nice if you move it up.

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:60]: we have a hook called `pythonRelaxDepsHook` which you should use instead of this. grep through nixpkgs for examples

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:61]: you should add: `doCheck = true;` since upstream has tests

File: pkgs/development/python-modules/anchor-kr/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anchor-kr/default.nix:1]: please also remove python3Packages here

File: pkgs/development/python-modules/anchor-kr/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anchor-kr/default.nix:6]:
```suggestion
buildPythonPackage {
```

File: pkgs/development/python-modules/anchor-kr/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anchor-kr/default.nix:18]: please also move format up

File: pkgs/development/python-modules/anitopy/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anitopy/default.nix:6]:
```suggestion
buildPythonPackage rec {
```

File: pkgs/development/python-modules/anitopy/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anitopy/default.nix:1]: please also remove python3Packages here

File: pkgs/development/python-modules/anitopy/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anitopy/default.nix:17]: same as with the other two please move it up to version

File: pkgs/development/python-modules/anitopy/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/anitopy/default.nix:20]: this package also has tests so you can also enable checks here with `doCheck = true;`

**Review by Janik-Haag — COMMENTED** [review]
Sorry that it took so long, I kinda forgot about this PR :see_no_evil:

**Comment by PassiveLemon** [comment]
Updated with changes

**Comment by PassiveLemon** [comment]
It now builds fine without needing to mess with the dependencies so I will remove those bits.

File: pkgs/applications/video/animdl/default.nix
Comment (Janik-Haag) [inline pkgs/applications/video/animdl/default.nix:4]:
```suggestion
  poetry-core,
  anchor-kr,
  anitopy,
```

**Review by Janik-Haag — COMMENTED** [review]
looks good to me except for this two missing attributes. @NobbZ can you take another look before we can hopefully get this merged in the next few days?

File: pkgs/applications/video/animdl/default.nix
Comment (PassiveLemon) [reply pkgs/applications/video/animdl/default.nix:4]: ah shoot forgot to put those back

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 240329` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>animdl</li>
    <li>animdl.dist</li>
    <li>python310Packages.anchor-kr</li>
    <li>python310Packages.anchor-kr.dist</li>
    <li>python310Packages.anitopy</li>
    <li>python310Packages.anitopy.dist</li>
    <li>python311Packages.anchor-kr</li>
    <li>python311Packages.anchor-kr.dist</li>
    <li>python311Packages.anitopy</li>
    <li>python311Packages.anitopy.dist</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]

**Comment by PassiveLemon** [comment]
I'm seeing some new change where packages are being located by name. Would you like me to update this PR?

**Comment by Janik-Haag** [comment]
Nah it's fine if you rebase, there will be a pr moving all the packages that can be moved to pkgs/by-name with a script, no need to do it manually. Sorry I forgot to merge this PR :see_no_evil:

**Comment by PassiveLemon** [comment]
Okay, I will rebase then

**Comment by PassiveLemon** [comment]
Done

**Comment by Janik-Haag** [comment]
@NobbZ can you also give you review again, merge is currently blocked because you requested some changes

**Comment by NobbZ** [comment]
Result of `nixpkgs-review pr 240329` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>animdl</li>
    <li>animdl.dist</li>
    <li>python310Packages.anchor-kr</li>
    <li>python310Packages.anchor-kr.dist</li>
    <li>python310Packages.anitopy</li>
    <li>python310Packages.anitopy.dist</li>
    <li>python311Packages.anchor-kr</li>
    <li>python311Packages.anchor-kr.dist</li>
    <li>python311Packages.anitopy</li>
    <li>python311Packages.anitopy.dist</li>
  </ul>
</details>

**Comment by NobbZ** [comment]
Result of `nixpkgs-review pr 240329` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python310Packages.anchor-kr</li>
    <li>python310Packages.anchor-kr.dist</li>
    <li>python310Packages.anitopy</li>
    <li>python310Packages.anitopy.dist</li>
    <li>python311Packages.anchor-kr</li>
    <li>python311Packages.anchor-kr.dist</li>
    <li>python311Packages.anitopy</li>
    <li>python311Packages.anitopy.dist</li>
  </ul>
</details>

**Review by NobbZ — APPROVED** [review]

**Pull request merged**
